diff --git a/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp b/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
index 0f2b969fba..ffd52644ca 100644
--- a/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
+++ b/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp
@@ -4447,7 +4447,11 @@ int AArch64InstrInfo::getMemScale(unsigned Opc) {
   case AArch64::STRWpre:
   case AArch64::LDPSi:
   case AArch64::LDPSWi:
+  case AArch64::LDPSWpre:
+  case AArch64::LDPSWpost:
   case AArch64::LDPWi:
+  case AArch64::LDPWpre:
+  case AArch64::LDPWpost:
   case AArch64::STPSi:
   case AArch64::STPWi:
     return 4;
@@ -4465,6 +4469,8 @@ int AArch64InstrInfo::getMemScale(unsigned Opc) {
   case AArch64::STRXpre:
   case AArch64::LDPDi:
   case AArch64::LDPXi:
+  case AArch64::LDPXpre:
+  case AArch64::LDPXpost:
   case AArch64::STPDi:
   case AArch64::STPXi:
     return 8;
diff --git a/llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp b/llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp
index cd976790eb..0b1543516d 100644
--- a/llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp
+++ b/llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp
@@ -177,7 +177,8 @@ struct AArch64LoadStoreOpt : public MachineFunctionPass {
   // pre or post indexed addressing with writeback. Scan forwards.
   MachineBasicBlock::iterator
   findMatchingUpdateInsnForward(MachineBasicBlock::iterator I,
-                                int UnscaledOffset, unsigned Limit);
+                                int UnscaledOffset, unsigned Limit,
+                                bool IsLDPPostIndex);
 
   // Scan the instruction list to find a register assigned with a const
   // value that can be combined with the current instruction (a load or store)
@@ -204,11 +205,17 @@ struct AArch64LoadStoreOpt : public MachineFunctionPass {
   bool isMatchingMovConstInsn(MachineInstr &MemMI, MachineInstr &MI,
                               unsigned IndexReg, unsigned &Offset);
 
+  enum LdStMergeKind {
+    PreIdx,
+    PostIdx,
+    LDPPostIdx,
+  };
+
   // Merge a pre- or post-index base register update into a ld/st instruction.
   std::optional<MachineBasicBlock::iterator>
   mergeUpdateInsn(MachineBasicBlock::iterator I,
                   MachineBasicBlock::iterator Update, bool IsForward,
-                  bool IsPreIdx, bool MergeEither);
+                  LdStMergeKind MergeKind, bool MergeEither);
 
   MachineBasicBlock::iterator
   mergeConstOffsetInsn(MachineBasicBlock::iterator I,
@@ -753,6 +760,7 @@ static bool isMergeableLdStUpdate(MachineInstr &MI, AArch64FunctionInfo &AFI) {
   case AArch64::LDRWui:
   case AArch64::LDRHHui:
   case AArch64::LDRBBui:
+  case AArch64::LDRSWui:
   case AArch64::STGi:
   case AArch64::STZGi:
   case AArch64::ST2Gi:
@@ -2099,7 +2107,7 @@ maybeMoveCFI(MachineInstr &MI, MachineBasicBlock::iterator MaybeCFI) {
 
 std::optional<MachineBasicBlock::iterator> AArch64LoadStoreOpt::mergeUpdateInsn(
     MachineBasicBlock::iterator I, MachineBasicBlock::iterator Update,
-    bool IsForward, bool IsPreIdx, bool MergeEither) {
+    bool IsForward, LdStMergeKind MergeKind, bool MergeEither) {
   assert((Update->getOpcode() == AArch64::ADDXri ||
           Update->getOpcode() == AArch64::SUBXri) &&
          "Unexpected base register update instruction to merge!");
@@ -2112,7 +2120,7 @@ std::optional<MachineBasicBlock::iterator> AArch64LoadStoreOpt::mergeUpdateInsn(
   // to do so, bail.
   MachineBasicBlock::iterator InsertPt = I;
   if (IsForward) {
-    assert(IsPreIdx);
+    assert(MergeKind == PreIdx);
     if (auto CFI = maybeMoveCFI(*Update, next_nodbg(Update, E)); CFI != E) {
       if (MergeEither) {
         InsertPt = Update;
@@ -2141,12 +2149,42 @@ std::optional<MachineBasicBlock::iterator> AArch64LoadStoreOpt::mergeUpdateInsn(
   if (Update->getOpcode() == AArch64::SUBXri)
     Value = -Value;
 
-  unsigned NewOpc = IsPreIdx ? getPreIndexedOpcode(I->getOpcode())
-                             : getPostIndexedOpcode(I->getOpcode());
+  unsigned NewOpc;
+  switch (MergeKind) {
+  case PreIdx:
+    NewOpc = getPreIndexedOpcode(I->getOpcode());
+    break;
+  case PostIdx:
+    NewOpc = getPostIndexedOpcode(I->getOpcode());
+    break;
+  case LDPPostIdx:
+    NewOpc = getPostIndexedOpcode(getMatchingPairOpcode(I->getOpcode()));
+    break;
+  }
   MachineInstrBuilder MIB;
   int Scale, MinOffset, MaxOffset;
   getPrePostIndexedMemOpInfo(*I, Scale, MinOffset, MaxOffset);
-  if (!AArch64InstrInfo::isPairedLdSt(*I)) {
+  if (MergeKind == LDPPostIdx) {
+    // Non-paired instruction transformed into a paired instruction.
+    assert((I->getOpcode() == AArch64::LDRXui ||
+            I->getOpcode() == AArch64::LDRWui ||
+            I->getOpcode() == AArch64::LDRSWui) &&
+           "Expected LDRXui or LDRWui instruction");
+    unsigned ZeroReg =
+        I->getOpcode() == AArch64::LDRWui ? AArch64::WZR : AArch64::XZR;
+    // NOTE: The scale wasn't calculated right above, so we need to update it.
+    Scale = TII->getMemScale(NewOpc);
+    MIB = BuildMI(*InsertPt->getParent(), InsertPt, InsertPt->getDebugLoc(),
+                  TII->get(NewOpc))
+              .add(Update->getOperand(0))
+              .addDef(ZeroReg)
+              .setOperandDead(1)
+              .add(getLdStRegOp(*I))
+              .add(AArch64InstrInfo::getLdStBaseOp(*I))
+              .addImm(Value / Scale)
+              .setMemRefs(I->memoperands())
+              .setMIFlags(I->mergeFlagsWith(*Update));
+  } else if (!AArch64InstrInfo::isPairedLdSt(*I)) {
     // Non-paired instruction.
     MIB = BuildMI(*InsertPt->getParent(), InsertPt, InsertPt->getDebugLoc(),
                   TII->get(NewOpc))
@@ -2169,7 +2207,7 @@ std::optional<MachineBasicBlock::iterator> AArch64LoadStoreOpt::mergeUpdateInsn(
               .setMIFlags(I->mergeFlagsWith(*Update));
   }
 
-  if (IsPreIdx) {
+  if (MergeKind == PreIdx) {
     ++NumPreFolded;
     LLVM_DEBUG(dbgs() << "Creating pre-indexed load/store.");
   } else {
@@ -2326,7 +2364,8 @@ bool AArch64LoadStoreOpt::isMatchingMovConstInsn(MachineInstr &MemMI,
 }
 
 MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnForward(
-    MachineBasicBlock::iterator I, int UnscaledOffset, unsigned Limit) {
+    MachineBasicBlock::iterator I, int UnscaledOffset, unsigned Limit,
+    bool IsLDPPostIndex) {
   MachineBasicBlock::iterator E = I->getParent()->end();
   MachineInstr &MemMI = *I;
   MachineBasicBlock::iterator MBBI = I;
@@ -2383,7 +2422,8 @@ MachineBasicBlock::iterator AArch64LoadStoreOpt::findMatchingUpdateInsnForward(
       ++Count;
 
     // If we found a match, return it.
-    if (isMatchingUpdateInsn(*I, MI, BaseReg, UnscaledOffset))
+    if (isMatchingUpdateInsn(*I, MI, BaseReg,
+                             IsLDPPostIndex ? 0 : UnscaledOffset))
       return MBBI;
 
     // Update the status of what the instruction clobbered and used.
@@ -2695,17 +2735,41 @@ bool AArch64LoadStoreOpt::tryToMergeLdStUpdate
   // add x20, x20, #32
   //   merged into:
   // ldr x0, [x20], #32
-  Update = findMatchingUpdateInsnForward(MBBI, 0, UpdateLimit);
+  Update = findMatchingUpdateInsnForward(MBBI, 0, UpdateLimit, false);
   if (Update != E) {
     // Merge the update into the ld/st.
-    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/false,
-                                     /*IsPreIdx=*/false,
+    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/false, PostIdx,
                                      /*MergeEither=*/false)) {
       MBBI = *NextI;
       return true;
     }
   }
 
+  // Look forward to try to form a post-index instruction using LDP.
+  // For example,
+  // ldr x0, [sp, #8]
+  // add sp, sp, #32
+  //   merged into:
+  // ldp xzr, x0, [sp], #32
+  if (MI.getOpcode() == AArch64::LDRWui || MI.getOpcode() == AArch64::LDRXui ||
+      MI.getOpcode() == AArch64::LDRSWui) {
+    // This transformation is only valid when we're loading from SP, because of
+    // its 16-byte alignment requirement.
+    if (MI.getOperand(1).isReg() && MI.getOperand(1).getReg() == AArch64::SP) {
+      Update = findMatchingUpdateInsnForward(MBBI, TII->getMemScale(*MBBI),
+                                             UpdateLimit, true);
+      if (Update != E) {
+        // Merge the update into the ld.
+        if (auto NextI =
+                mergeUpdateInsn(MBBI, Update, /*IsForward=*/false, LDPPostIdx,
+                                /*MergeEither=*/false)) {
+          MBBI = *NextI;
+          return true;
+        }
+      }
+    }
+  }
+
   // Don't know how to handle unscaled pre/post-index versions below, so bail.
   if (TII->hasUnscaledLdStOffset(MI.getOpcode()))
     return false;
@@ -2719,8 +2783,8 @@ bool AArch64LoadStoreOpt::tryToMergeLdStUpdate
   Update = findMatchingUpdateInsnBackward(MBBI, UpdateLimit, MergeEither);
   if (Update != E) {
     // Merge the update into the ld/st.
-    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/true,
-                                     /*IsPreIdx=*/true, MergeEither)) {
+    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/true, PreIdx,
+                                     MergeEither)) {
       MBBI = *NextI;
       return true;
     }
@@ -2737,11 +2801,11 @@ bool AArch64LoadStoreOpt::tryToMergeLdStUpdate
   // add x0, x0, #64
   //   merged into:
   // ldr x1, [x0, #64]!
-  Update = findMatchingUpdateInsnForward(MBBI, UnscaledOffset, UpdateLimit);
+  Update =
+      findMatchingUpdateInsnForward(MBBI, UnscaledOffset, UpdateLimit, false);
   if (Update != E) {
     // Merge the update into the ld/st.
-    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/false,
-                                     /*IsPreIdx=*/true,
+    if (auto NextI = mergeUpdateInsn(MBBI, Update, /*IsForward=*/false, PreIdx,
                                      /*MergeEither=*/false)) {
       MBBI = *NextI;
       return true;
